planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+  different kinds of streams
+* Postgresql native partitioning by tenant id where that's useful
+* Aggregation keys -- this is tightly coupled to the projections. If there's a projection that is done by something other than tenant id or stream id, try to persist the 
+  key for however the events/streams are grouped to stream and/or event storage. Thinking about data elements like "Region" or "Country" where you might be doing
+  aggregations across streams or tenants. Maybe that's useful?
+* Archive streams that reach some kind of termination event? Move everything to an archive table off to the side? Keeps the event log table cleaner
+* Artificial segmenting of the streams. Use some kind of algorithm that could see stream ids, and assign them to 2 or more "cohorts", then persist that "cohort key" to 
+  both the stream and event data. This would potentially help for the async daemon to parallelize work. It would definitely help out for "live" projections by having 
+  smaller tables to work through
+
+### Projections
+
+Thoughts:
+
+Functionality wise, the lifecycle of projections are:
+
+* "Live" projections built on demand from the raw events. This only changes for v4 to take advantage of new snapshotting capabilities
+* "Inline" projections built as part of the same unit of work as capturing the source events. See the section on *Inline Projections* because this
+  will change internally for v4
+* "Async" projections built up in a background process of some sort. I think this turns into a near re-write for v4
+
+By *type*, the projection types *could* be:
+
+* Aggregate document by stream (existing)
+* Aggregate document by tenant id (new)
+* Execute SQL statement by event type (new) -- so, insert/update/delete. This comes up a lot, and it should be low hanging fruit
+* Transform a single event to a document --> needs to become just "carry out some kind of change to a document" and probably ends up sharing more mechanics with the aggregation than it used to
+* Aggregate by arbitrary values of either the stream or event types -- but if we do this, I strongly vote to persist whatever the differentiation value is in the database to make projection rebuilds and the async daemon more "parallelizable"
+
+
+Other thoughts:
+
+* I want to completely throw away the current `Aggregator` implementations. It was just too naive for real world usage and the internals have gotten ugly
+* Re-evaluate the `ViewProjection` fluent interface. That might fit in fine to our new model of projections for v4 with more overloads
+* I've been in favor of having adapters for existing projection libraries like [Liquid Projections](https://liquidprojections.net/) in the past, but I think we'd be losing too
+  much optization, especially for the aggregations inside the async daemon.
+
+#### Defining a Projection to a Document
+
+
+```
+ProjectTo<ProjectedDocument>()
```

Comment:
I'm not sure how our users are using that, but I've never used this fluent syntax in my previous project. I always created dedicated ViewProjection class. It was for me much easier to maintain.

Reply:
Using object scoping instead? I'm cool with that.

Reply:
I agree with @oskardudycz. The fluent syntax looks nice but projection logic is rarely so "small" that it feels like it fits well in this place. Id rather have it in a separate class. Also, with some improvements to the design of the projection classes that would also make it easier to unit test projections (which sadly is a bit too messy today).

Reply:
Yup, unit testing is a valid case. I also heard from few people that "Marten is not easily unit testable". However, I don't fully agree with the statement, but projections for sure should be possible to be tested without Marten infrastructure.

Reply:
"Marten is not easily unit testable". -- because folks shoot themselves in the foot by trying to mock out the Linq support. I still need to do some testing recommendations. 

And I think your code is structured poorly if you think you need to mock Marten to test your business functionality.

Reply:
@jeremydmiller I fully agree for that point. But still, it'd be good to verify for the projection if it's mapping the fields properly without the need to call DB.